### PR TITLE
Make `rasa visualize` usable for Deutsche Telekom

### PR DESCRIPTION
**Pull request by raoulvm**
**Proposed changes**:
- Visualize **slots** with values 
  - stories might branch at slots being set (SlotSet event)
- Visualize **entities** in intents 
  - stories might be different depending on the existence of entities in the intent (UserUttered event)
- Instead of picking random training examples, print the **intent name**

Follow up to 
* https://github.com/RasaHQ/rasa/issues/10941
* https://github.com/RasaHQ/rasa/pull/10932 
 
  See **example pictures** there. Also there is a comment from @amn41  in https://github.com/RasaHQ/rasa/issues/10941#issuecomment-1152187472

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by raoulvm** [comment]
@souvikg10 This is the PR.

**Review by Urkem — CHANGES_REQUESTED** [review]
Thanks for the contribution 💯. I've added some comments below. Can you please add some tests or share an example to verify that it works?

File: rasa/shared/core/training_data/visualization.py [hunk 1]
```diff
@@ -517,7 +570,9 @@ def visualize_neighborhood(
 
     if should_merge_nodes:
         _merge_equivalent_nodes(graph, max_history)
-    _replace_edge_labels_with_nodes(graph, next_node_idx, nlu_training_data)
+    _replace_edge_labels_with_nodes(
+        graph, next_node_idx, None
+    )  # None instead of nlu_training_data)
```
[the remark this answers is not in the document]
Comment (raoulvm) [reply rasa/shared/core/training_data/visualization.py:575]: If you pass nlu_training_data the intent name will be replaced by some random user example. The diagram becomes mostly unreadable with that. If you have an idea how to make that behavior parametrizable be my guest.

File: rasa/shared/core/training_data/visualization.py [hunk 2]
```diff
@@ -454,15 +478,44 @@ def visualize_neighborhood(
...
+                    next_node_idx += 1
+                    graph.add_node(
+                        next_node_idx,
+                        label=label,
+                        fontsize=fontsize,
+                        style="filled",
+                        **{"class": "slot"},
+                    )
+
+                    _add_message_edge(
```
[the remark this answers is not in the document]
Comment (raoulvm) [reply rasa/shared/core/training_data/visualization.py:505]: Oh, I copied that from here: 
https://github.com/RasaHQ/rasa/pull/12387/files/687a06a8b4311707017febdb73adccdf72df483a#diff-f1f79fa3357903f6df629d9baa2e523887e92da6e29aaf820231b94d57a777ddR523-R525

File: rasa/shared/core/training_data/visualization.py [hunk 3]
```diff
@@ -403,19 +407,33 @@ def _add_message_edge(
 ) -> None:
     """Create an edge based on the user message."""
 
-    if message:
-        message_key = message.get("intent", {}).get("name", None)
+    if message and message.get("intent", {}).get("name", None):
```
Comment (ancalita) [inline rasa/shared/core/training_data/visualization.py:410]: I'd suggest extracting `message.get("intent", {}).get("name", None)` to a different variable outside the if/elif block since you're using it twice in this if branch (here and in L421)

This way you can also keep message_key = <the_new_variable> in L426 instead if assigning it to `message.label`.

**Comment by souvikg10** [comment]
@ancalita @Urkem - I have modified the tests and fixed the lint and types. It is ready for review.

File: rasa/shared/core/training_data/visualization.html [hunk 4]
```diff
@@ -8,139 +9,147 @@
...
-    serveGraph();
-  }
-
-
-</script>
+    </script>
 </body>
-</html>
+
+</html>
```
Comment (ancalita) [inline rasa/shared/core/training_data/visualization.html:155]: Missing new line.

File: rasa/shared/core/training_data/visualization.py [hunk 5]
```diff
@@ -402,20 +400,40 @@ def _add_message_edge(
...
+                if key == "entity"
+            ]
+        if entity_values:
+            newline = ",\n"
+            entity_values_as_str = f"{newline}({newline.join(entity_values)})"
+        else:
+            entity_values_as_str = ""
+
+        message_key = message_intent + entity_values_as_str
+        message_label = message_key  # message.get("text", None)
```
Comment (ancalita) [inline rasa/shared/core/training_data/visualization.py:422]: What does this comment represent? Is it supposed to be removed?

File: rasa/shared/core/training_data/visualization.py [hunk 6]
```diff
@@ -402,20 +400,40 @@ def _add_message_edge(
...
+    elif message and message.get("text"):
+
         message_label = message.get("text", None)
+        message_key = message.get("intent", {}).get("name", None)
```
Comment (ancalita) [inline rasa/shared/core/training_data/visualization.py:426]: I don't think this will ever get evaluated because it's a subset condition of first if branch condition,  unless you specify in L403: `if message and message.get("text") is None`.

File: rasa/shared/core/training_data/visualization.py [hunk 7]
```diff
@@ -454,15 +478,44 @@ def visualize_neighborhood(
                 idx -= 1
                 break
             if isinstance(el, UserUttered):
+                if message is not None:
+                    logger.error("multiple messages or slots in a row!!")
```
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/shared/core/training_data/visualization.py:482]: This still needs to be addressed with indication on how to resolve the error. Also not sure if logging is appropriate here if the error lies with the user sending multiple messages?

File: rasa/shared/core/training_data/visualization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/shared/core/training_data/visualization.py:575]: We need an improved comment describing concisely why None is required here (as explained above).

File: tests/shared/core/training_data/test_visualization.py [hunk 8]
```diff
@@ -142,9 +142,9 @@ def test_story_visualization(domain: Domain, tmp_path: Path):
     )
 
     assert str(None) not in out_file.read_text()
-    assert "/affirm" in out_file.read_text()
-    assert len(generated_graph.nodes()) == 51
-    assert len(generated_graph.edges()) == 56
+    assert "affirm" in out_file.read_text()
+    assert len(generated_graph.nodes()) == 53
+    assert len(generated_graph.edges()) == 58
```
Comment (ancalita) [inline tests/shared/core/training_data/test_visualization.py:147]: I'm not familiar with these unit tests, so a quick explanation on why the changes in affected unit tests were required would be helpful 🙏🏻

**Review by ancalita — COMMENTED** [review]
Nice work 💯 ! 
@souvikg10 left some additional comments, to double-check was this manually tested in a local environment that contains these changes?

File: rasa/shared/core/training_data/visualization.py (hunk 5 above)
Comment (souvikg10) [reply rasa/shared/core/training_data/visualization.py:422]: I just realised, what was the need here, i believe this is to make sure that the label of the edge, shows the intent and entity values as str, but according to me, it is fine using the text as label as well, reverted this.

File: rasa/shared/core/training_data/visualization.py (hunk 6 above)
Comment (souvikg10) [reply rasa/shared/core/training_data/visualization.py:426]: Yeah, i updated the first condition, so now, we have two conditions,
if message has entities
then first condition
else if message has text
then second condition

File: tests/shared/core/training_data/test_visualization.py (hunk 8 above)
Comment (souvikg10) [reply tests/shared/core/training_data/test_visualization.py:147]: For the tests, it is quite simple, we have new graph nodes and edges now because we are adding slots onto the visualization, which wasn't the case previously.

File: rasa/shared/core/training_data/visualization.py (hunk 5 above)
Comment (souvikg10) [reply rasa/shared/core/training_data/visualization.py]: i suppose i understand, why seeing the entities would make sense here. removed the comment

File: rasa/shared/core/training_data/visualization.py [hunk 9]
```diff
@@ -456,6 +479,33 @@ def visualize_neighborhood(
             if isinstance(el, UserUttered):
                 message = el.parse_data
```
Comment (souvikg10) [inline rasa/shared/core/training_data/visualization.py:480]: i don't think the check made any sense there. i don't when in UserUttered, you would have multiple slots in a row. That's not the same event.

File: rasa/shared/core/training_data/visualization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (souvikg10) [reply rasa/shared/core/training_data/visualization.py]: @ancalita @raoulvm - I think you have two usecases here, 
If you are passing nlu_training_data, it takes a random example from the training set, which i believe is fine. I wouldn't pass None here because then userMessage is never shown. There is already a check in the method to see when nlu_training_data is None
If you only want to see the intents, then don't pass the nlu training data, it is already an optional param in the cli.

File: rasa/shared/core/training_data/visualization.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (souvikg10) [reply rasa/shared/core/training_data/visualization.py]: As visualize is taken from the training dataset, i don't think there is such a case in the events where a UserUttered is multiple messages in a row. regarding multiple slots, it would be taken from SlotSet event and not UserUttered. This check is not needed imo @ancalita

File: rasa/shared/core/training_data/visualization.py (hunk 5 above)
Comment (raoulvm) [reply rasa/shared/core/training_data/visualization.py]: Thanks, I was just about to comment on the comment. It is not the entity values, it is the entity keys, maybe renaming the variable would help here.

File: rasa/shared/core/training_data/visualization.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (raoulvm) [reply rasa/shared/core/training_data/visualization.py]: It can only happen if you have an error in the training stories, like two consecutive intents.

[the remark this answers is not in the document]
Comment (raoulvm) [reply rasa/shared/core/training_data/visualization.py]: When I developed the code last year, I also tried to break it. That was one way to make the python fail or make a nonsense diagram without any warning, so I added the (admittedly very short) error message here.

File: rasa/shared/core/training_data/visualization.py [hunk 10]
```diff
@@ -42,6 +42,8 @@
 
 VISUALIZATION_TEMPLATE_PATH = "/visualization.html"
 
+logger = logging.getLogger(__name__)
```
Comment (Urkem) [inline rasa/shared/core/training_data/visualization.py:45]: I don't see the `logger` being used 🤔 can you remove it? (or add some debug, info etc. statements)

**Review by Urkem — COMMENTED** [review]
Just small suggestions 👍
